### PR TITLE
[FIRRTL] Eliminate FlipType

### PR DESCRIPTION
**Pull request by darthscsi**
FlipType doesn't propagate the same way Flow does, and becomes a redundant way to try to represent FIRRTL's flow concept.  This matches FIRRTL by attaching flip as a property of bundle fields.  This eliminates a fair amount of code which constantly was looking around flips.  Only two operations need special consideration: InstanceOp and MemOp.  When computing flow for an InstanceOp result, you need to look at whether it was an input or output for the referenced module.  For MemOp, all flows are sinks.

### CODE REVIEW

**Review by mikeurbach — APPROVED** [review]
This is awesome! Thanks for dealing with the Handshake stuff. I only reviewed those changes but they look good to me. I'll double check the Handshake integration tests after this lands.

**Comment by lattner** [comment]
I love it, so awesome.  Thank you!!

**Review by lattner — APPROVED** [review]

File: test/CAPI/CMakeLists.txt [hunk 1]
```diff
@@ -8,7 +8,7 @@ target_link_libraries(circt-capi-ir-test
   PRIVATE
   ${dialect_libs}
 
-  MLIRCAPIRegistration
+#  MLIRCAPIRegistration
```
Comment (seldridge) [inline test/CAPI/CMakeLists.txt:11]: I think this was a stray comment that got in here?

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 2]
```diff
@@ -116,8 +116,15 @@ Flow firrtl::foldFlow(Value val, Flow accumulatedFlow) {
       // Registers, Wires, and behavioral memory ports are always Duplex.
       .Case<RegOp, RegResetOp, WireOp, MemoryPortOp>(
           [](auto) { return Flow::Duplex; })
-      .Case<InstanceOp>([&](auto) {
-        return val.getType().isa<FlipType>() ? swap() : accumulatedFlow;
+      .Case<InstanceOp>([&](auto inst) {
+        for (auto arg : llvm::enumerate(inst.getResults()))
+          if (arg.value() == val)
+            if (getModulePortDirection(inst.getReferencedModule(),
+                                       arg.index()) == Direction::Output)
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:123]: Is `getModulePortDirection` safe to use in a parallel context where modules are being operated on independently? (E.g., in LowerTypes?)

**Review by seldridge — COMMENTED** [review]
This is awesome, @darthscsi. Flip is _finally_ gone.

Thanks for taking the time to do this. This was risking a repetitive strain injury just to update the tests. 😬

**Pull request merged**